system-addon/content-src/components/TopSites/TopSites.jsx

```diff
@@ -331,14 +371,194 @@ class TopSitesEdit extends React.Component {
+            <TopSiteForm
+              label={this.props.TopSites.rows[this.state.editIndex].label || this.props.TopSites.rows[this.state.editIndex].hostname}
+              url={this.props.TopSites.rows[this.state.editIndex].url}
+              index={this.state.editIndex}
+              editMode={true}
+              onClose={this.onFormClose}
+              dispatch={this.props.dispatch}
+              intl={this.props.intl} />
+          </div>
+        </div>
+      }
     </div>);
   }
 }
 
 const TopSitesEditIntl = injectIntl(TopSitesEdit);
 
+class TopSiteForm extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      label: props.label || "",
+      url: props.url || "",
+      validationError: false
+    };
+    this.onLabelChange = this.onLabelChange.bind(this);
+    this.onUrlChange = this.onUrlChange.bind(this);
+    this.onCancelButtonClick = this.onCancelButtonClick.bind(this);
+    this.onAddButtonClick = this.onAddButtonClick.bind(this);
+    this.onSaveButtonClick = this.onSaveButtonClick.bind(this);
+    this.onUrlInputMount = this.onUrlInputMount.bind(this);
+  }
+  onLabelChange(event) {
+    this.resetValidation();
+    this.setState({"label": event.target.value});
+  }
+  onUrlChange(event) {
+    this.resetValidation();
+    this.setState({"url": event.target.value});
+  }
+  onCancelButtonClick(ev) {
+    ev.preventDefault();
+    this.props.onClose();
+  }
+  onAddButtonClick(ev) {
+    ev.preventDefault();
+    if (this.validateForm()) {
+      let site = {url: this.cleanUrl()};
+      if (this.state.label !== "") {
+        site.label = this.state.label;
+      }
+      this.props.dispatch(ac.SendToMain({
+        type: at.TOP_SITES_ADD,
+        data: {site}
+      }));
+      this.props.dispatch(ac.UserEvent({
+        source: TOP_SITES_SOURCE,
+        event: "TOP_SITES_ADD"
+      }));
+      this.props.onClose();
+    }
+  }
+  onSaveButtonClick(ev) {
+    ev.preventDefault();
+    if (this.validateForm()) {
+      let site = {url: this.cleanUrl()};
+      if (this.state.label !== "") {
+        site.label = this.state.label;
+      }
+      this.props.dispatch(ac.SendToMain({
+        type: at.TOP_SITES_EDIT,
+        data: {site, index: this.props.index}
+      }));
+      this.props.dispatch(ac.UserEvent({
+        source: TOP_SITES_SOURCE,
+        event: "TOP_SITES_EDIT",
+        action_position: this.props.index
+      }));
+      this.props.onClose();
+    }
+  }
+  cleanUrl() {
+    let url = this.state.url;
+    // If we are missing a protocol, prepend http://
+    if (!url.startsWith("http")) {
+      url = `http://${url}`;
+    }
+    return url;
+  }
+  resetValidation() {
+    if (this.state.validationError) {
+      this.setState({validationError: false});
+    }
+  }
+  validateUrl() {
+    const pattern = new RegExp("^(https?:\\/\\/)?" + // protocol
+      "((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|" + // domain name
```

Comment:
Hmmm, so if I edit facebook.com to somethingelse.com yeah it basically creates a new site in that spot so facebook.com would probably appear next to it after. Not sure if we can change that other than blocking it and that's probably not right. I guess to me edit means that I want to edit what appears in this square and not that I want to edit the site data itself (unless the url stays the same).

It's the same behavior that we had in test pilot so we should probably leave it like as it was?

Reply:
So I guess the question is does edit mean:
a) "Edit this site, but also keep the original" 
b) "Edit this site, and discard what I had originally"

Which is a design question I suppose. Yeah I think it's fine to leave this like this. In the worst case they can always block the site they no longer want anyways

Reply:
btw, i hate this regex. would love to kill it if something similar already exists in m-c

Reply:
Not sure if we can just rely on `new URL()` and additionally restrict to http/s

Reply:
Oh that's an interesting idea. The only thing is we want the http://|https:// to be optional. But we could probably prepend that when it's needed before feeding it into `URL()`.

Reply:
(I'll play with it and propose something in a new PR 👍 )

Reply:
Stuff in mozilla-central uses [nsIURIFixup](http://searchfox.org/mozilla-central/source/docshell/base/nsIURIFixup.idl) in various ways; if you walk through [the urlbar code itself](http://searchfox.org/mozilla-central/source/browser/base/content/urlbarBindings.xml) searching for the string "fixup", you'll see a variety of uses.

We also had something like this in Hello; it's also a very nasty regexp, but it's relatively well-tested, and also tries to handle some security issues.  That said, the purpose of the code isn't quite the same, so the security issues and edge-cases here might be different.  

https://github.com/mozilla/loop/blob/master/shared/js/urlRegExps.js
https://github.com/mozilla/loop/blob/master/shared/js/linkifiedTextView.jsx
https://github.com/mozilla/loop/blob/master/shared/test/linkifiedTextView_test.js

Whether it would actually be a net benefit to switch to the Hello code, I can't guess.

Reply:
Also, there are some tools for regexp debugging, should you still need it: https://wiki.mozilla.org/Loop/Development/RegExpDebugging

Reply:
Thanks @dmose! We ended up just using `new URL()` as a validator 560e9b62a2bde24feb795f16418ffd49df54ee95. It's good enough for what we need. I bookmarked those regex tools as I'm sure I'll need them again sometime.